flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/TableMetadataCache.java

```diff
@@ -194,63 +208,62 @@ static class CacheItem {
 
     private final boolean tableExists;
     private final Set<String> branches;
-    private final SchemaInfo schema;
+    private final Map<Integer, Schema> tableSchemas;
     private final Map<Integer, PartitionSpec> specs;
+    private final LRUCache<Schema, ResolvedSchemaInfo> inputSchemas;
 
     private CacheItem(
         boolean tableExists,
         Set<String> branches,
-        SchemaInfo schema,
-        Map<Integer, PartitionSpec> specs) {
+        Map<Integer, Schema> tableSchemas,
+        Map<Integer, PartitionSpec> specs,
+        int inputSchemaCacheMaximumSize) {
       this.tableExists = tableExists;
       this.branches = branches;
-      this.schema = schema;
+      this.tableSchemas = tableSchemas;
       this.specs = specs;
+      this.inputSchemas =
+          new LRUCache<>(inputSchemaCacheMaximumSize, CacheItem::inputSchemaEvictionListener);
+    }
+
+    private static void inputSchemaEvictionListener(
+        Map.Entry<Schema, ResolvedSchemaInfo> evictedEntry) {
+      LOG.warn(
+          "Performance degraded as records with different schema is generated for the same table. "
+              + "Likely the DynamicRecord.schema is not reused. "
+              + "Reuse the same instance if the record schema is the same to improve performance");
     }
 
     @VisibleForTesting
-    SchemaInfo getSchemaInfo() {
-      return schema;
+    Map<Schema, ResolvedSchemaInfo> getInputSchemas() {
```

Comment:
Could we rename this method to `inputSchemas`. Sorry that I missed this: https://iceberg.apache.org/contribute/#method-naming